### DISCUSSION TITLE
Can someone help me understand the design considerations/requirements around the typing system?

### DISCUSSION THREAD

**Discussion by trevor-coleman** [js-fns/date-fns#3453] [Q&A]
**tl;dr I'm looking to open a PR to improve the way that type definitions are generated for the package, and looking to get some context on how/why the types are generated the way they are.**

## Background 
My team recently encountered an issue with `date-fns` and the `import/no-duplicates` rule in `eslint-plugin-import`, where imports like this are flagged as duplicates:

```
import addSeconds from 'date-fns/addSeconds'
import addMinutes from 'date-fns/addMinutes'
```

According to [this discussion](https://github.com/import-js/eslint-plugin-import/issues/1479) the reason they are flagged as duplicates is because all of the typescript types for the project are contained in `typings.d.ts` and declared as modules like this:

```
declare module 'date-fns/addSeconds/index' {
  import { addSeconds } from 'date-fns'
  export default addSeconds
}
```

This is a fairly old-school approach to types, and it would be good to modernize it, given how critical this library is. I see that the typescript migration started in 2020, and looks to be nearly complete, so I'd love to pitch in on this last part.

## How the Types are Built
I took a look and the types are generated in `~/scripts/build/_lib/typings/typeScript.js` ([here](https://github.com/date-fns/date-fns/blob/main/scripts/build/_lib/typings/typeScript.js))

It seems that the [build script](https://github.com/date-fns/date-fns/blob/main/scripts/build/build.sh) creates `docs.json` as a single source of truth for all types and then constructs the typescript types from that, outputting them all into `~/typings.d.ts`

## Questions

My main questions are: 

1. Given that the whole library is written in typescript, is there a reason why we aren't using the typescript compiler to generate the types? 
2. Are there important design or architectural considerations to keep in mind when working with this system that aren't obvious?
3. Is there any active work in this area I should be aware of, or that I could help with instead of opening my own PR?

##  Additional notes
-  I see that the script is also writing types for flow as well, and I understand that the library needs to work in pure-JS environments. Is there something about these requirements that prevents us from just using the TS types? Like could we use the existing script to generate flow types, and then just tsc for typescript?
- Is there something I'm missing about `docs.json` -- for instance are only a specific subset of the exports that are included? or could we keep using that for flow types and just use `tsc` to generate the TS types?

## Possible Approaches

I've been thinking a bit about this, and have thought of a couple approaches, and could use some feedback on which seems like the right way to go:

### 1. Fix-in-place
- just work with the existing build script to get it to output things in a different format. 
- probably with a separate file for each module, re-exported through typings.d.ts
- Pros: 
      - least amount of change
      - people already know how it works
      - keeps flow and TS together in one place
      - we don't move anything around
- Cons: 
      - the code is still in pure-JS so there are lots of places to introduce bugs
      - it's a bit indirect, we have types, then we use this complicated machine to remove them and rebuild them from scratch somewhere else
      
### 2. Use tsc to generate the types
- here we separate the TS and flow types, we use the typescript compiler to build the types
- we would set up the build process to output to a `/build` or `/lib` directory, with type definitions next to their js counterparts
- flow types would be generated using the existing system (or maybe that could also be updated? I would need to look into flow)
- update packageJSON to point to the new build folder as the output.
```
~
├── src
│   └── //...
└── build
    ├── addSeconds.js
    ├── addSectonds.d.ts
    ├── addMinutes.js
    └── addMinutes.d.ts
```

- Pros:
      - will be forward compatible, as tsc will always output types idiomatically
      - it's the standard way to process typescript
- Cons
      - changes where things live
      - flow and ts using different systems, could drift apart, create future maintenance issues
  
**Cleaner, more idiomatic. But potential to cause drift betwen flow to typescript.**
      
## Final thoughts

#2 seems like the winner to me, but this feels a bit like "fools rush in" territory, so I would love to get some input from people with more experience in the project before I get in too deep on it.
